crates/matrix-sdk-ui/src/authentication/oidc.rs

```diff
@@ -0,0 +1,283 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    collections::HashMap,
+    fmt, fs,
+    fs::File,
+    io::{BufReader, BufWriter},
+    path::PathBuf,
+};
+
+use matrix_sdk::oidc::types::registration::{
+    ClientMetadata, ClientMetadataVerificationError, VerifiedClientMetadata,
+};
+use serde::{Deserialize, Serialize};
+use url::Url;
+
+#[derive(Debug, thiserror::Error)]
+pub enum OidcRegistrationsError {
+    #[error("Failed to use the supplied base path.")]
+    InvalidBasePath,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum RegistrationDataError {
+    #[error("Failed to load registrations file: {message}")]
+    LoadFailure { message: String },
+    #[error("Metadata hash mismatch, clearing registrations")]
+    MetadataMismatch,
+}
+
+/// A client ID that has been registered with an OpenID Connect provider.
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct ClientId(pub String);
+
+impl fmt::Display for ClientId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
```

Comment:
I think this is used indirectly for a `to_string()` call happening in the FFI; instead, since the wrapped `String` is public, could you remove this implementation, so there's no chance to do an accidental allocation, and instead callers can directly use `.0`?